node/bft/src/helpers/cache_round.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright (C) 2019-2023 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::{ensure, Result};
+use snarkvm::{
+    console::types::{Address, Field},
+    ledger::committee::Committee,
+    prelude::Network,
+};
+use std::collections::HashSet;
+
+#[derive(Copy, Clone, Debug)]
+struct AddressWithCoordinate<N: Network> {
+    address: Address<N>,
+    x: Field<N>,
+}
+
+impl<N: Network> From<Address<N>> for AddressWithCoordinate<N> {
+    fn from(address: Address<N>) -> Self {
+        Self { address, x: address.to_group().to_x_coordinate() }
+    }
+}
+
+#[derive(Debug)]
+pub struct RoundCache<N: Network> {
+    /// The current highest round which has (stake-weighted) quorum
+    last_highest_round_with_quorum: u64,
+    /// A sorted list of (round, Vec<AddressWithCoordinate<N>>), indicating the last seen highest round for each address
+    highest_rounds: Vec<(u64, Vec<AddressWithCoordinate<N>>)>,
```

Comment:
Why does this `Vec` have to be sorted? It seems only to quickly fetch an item? If that is true, why not use `HashMap`? This gives you O(1) lookups and O(1) inserts. Now you have O(n) inserts and O(log n) lookups.

Reply:
Will come back to this if we decide to keep this PR

Reply:
it's faster and more lightweight, unless the collection can become arbitrarily large